### PR TITLE
[docs] ❗️BREAKING ❗️ various theme config options was renamed

### PR DESCRIPTION
**Pull request by B2o5T**
fixes https://github.com/shuding/nextra/issues/676 https://github.com/shuding/nextra/issues/669

also, add new option `toc.component` to fix https://github.com/shuding/nextra/issues/564

and fixes `project.link` behaviour https://github.com/shuding/nextra/issues/383

https://github.com/shuding/nextra/blob/bbfe8a37ce59fc0bbb00da7c0d8a04220d111c53/packages/nextra-theme-docs/src/types.ts#L15-L82

@shuding it was hard to implement without too monkey-business code for some options like
```ts
 search: boolean | FC | ReactNode | {
   placeholder?: string | (() => string)
   emptyResult?: ReactNode | FC
 }
```

So I made the following

```ts
 search: { 
     component: ReactNode | FC<{ directories: Item[] }> 
     emptyResult: ReactNode | FC  // will be ignored if component was specified
     placeholder: string | (() => string)  // will be ignored if component was specified
   } 
```

or more complex usage

```ts
  toc: {
    component: ReactNode | FC<TOCProps>
    extraContent: ReactNode | FC // will be ignored if component was specified
    float: boolean // ❗️ headings will be passed in `toc.component` as props based on `headings: config.toc.float ? headings : []` condition
    title: ReactNode | FC // will be ignored if component was specified
  }
```

### CODE REVIEW

File: packages/nextra-theme-docs/src/contexts/config.tsx [hunk 1]
```diff
@@ -33,15 +74,48 @@ export const ConfigProvider = ({
...
+      const option = (themeConfig as any)[key]
+      if (typeof option === 'boolean' || option == null) {
+        console.warn(
+          `${notice} "${key}".`,
+          option ? 'Remove it' : `Rename it to ${key}: { component: null }`
+        )
+      }
+    }
+    if (typeof themeConfig.banner === 'string') {
+      console.warn(notice, '"banner". `Rename it to banner: { text: ... }')
```
Comment (B2o5T) [inline packages/nextra-theme-docs/src/contexts/config.tsx:116]:
```suggestion
      console.warn(notice, '"banner". Rename it to banner: { text: ... }')
```

File: packages/nextra-theme-docs/src/components/footer.tsx [hunk 2]
```diff
@@ -5,9 +5,8 @@ import { LocaleSwitch } from './locale-switch'
 import { ThemeSwitch } from './theme-switch'
 import { renderComponent } from '../utils'
 
-export function Footer({ menu }: { menu?: boolean }): ReactElement {
+export function Footer({ menu }: { menu?: boolean }): ReactElement | null {
```
Comment (B2o5T) [inline packages/nextra-theme-docs/src/components/footer.tsx:8]:
```suggestion
export function Footer({ menu }: { menu?: boolean }): ReactElement {
```

**Review by shuding — APPROVED** [review]
Thank you this is so much cleaner!

**Comment by Filsommer** [comment]
Regarding this change:
`placeholder: string | (() => string)`
How can we change that placeholder depending on the locale now?

**Pull request merged**